crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,732 @@
+        "Add missing blank line".to_string()
+    }
+}
+
+/// Returns `true` if the token is Async, Class or Def
+fn is_top_level_token(token: &Option<Tok>) -> bool {
+    matches!(&token, Some(Tok::Class | Tok::Def))
+}
+
+/// Returns `true` if the token is At, Async, Class or Def
+fn is_top_level_token_or_decorator(token: &Tok) -> bool {
+    matches!(&token, Tok::Class | Tok::Def | Tok::At)
+}
+
+#[derive(Debug)]
+struct LogicalLineInfo {
+    first_token: Tok,
+    first_token_range: TextRange,
+    last_token: Tok,
+    last_token_range: TextRange,
+    is_comment_only: bool,
+    is_docstring: bool,
+    indent_level: usize,
+    blank_lines: u32,
+    preceding_blank_lines: u32,
+    preceding_blank_characters: TextSize,
+}
+
+/// Iterator that processes tokens until a full logical line (or comment line) is "built".
+/// It then returns characteristics of that logical line (see `LogicalLineInfo`).
+struct LinePreprocessor<'a> {
+    tokens: Flatten<Iter<'a, Result<(Tok, TextRange), LexicalError>>>,
+    locator: &'a Locator<'a>,
+    /// Number of previous consecutive blank lines.
+    previous_blank_lines: u32,
+    /// Number of consecutive blank lines.
+    current_blank_lines: u32,
+    /// Number of blank characters in the blank lines (\n vs \r\n for example).
+    current_blank_characters: TextSize,
+}
+
+impl<'a> LinePreprocessor<'a> {
+    fn new(tokens: &'a [LexResult], locator: &'a Locator) -> LinePreprocessor<'a> {
+        LinePreprocessor {
+            tokens: tokens.iter().flatten(),
+            locator,
+            previous_blank_lines: 0,
+            current_blank_lines: 0,
+            current_blank_characters: 0.into(),
+        }
+    }
+}
+
+impl<'a> Iterator for LinePreprocessor<'a> {
+    type Item = LogicalLineInfo;
+
+    fn next(&mut self) -> Option<LogicalLineInfo> {
+        let mut line_is_comment_only = true;
+        let mut is_docstring = true;
+        let mut first_token: Option<Tok> = None;
+        let mut first_token_range: Option<TextRange> = None;
+        let mut last_token: Option<Tok> = None;
+        let mut parens = 0u32;
+
+        while let Some((token, range)) = self.tokens.next() {
+            if matches!(token, Tok::Indent | Tok::Dedent) {
+                continue;
+            }
+
+            if !matches!(token, Tok::Newline) {
+                // Ideally, we would like to have a "async def" token since we care about the "def" part.
+                // As a work around, we ignore the first token if it is "async".
+                if first_token.is_none() && !matches!(token, Tok::Async) {
+                    first_token = Some(token.clone());
+                }
+                if first_token_range.is_none() {
+                    first_token_range = Some(*range);
+                }
+
+                if !matches!(token, Tok::NonLogicalNewline) {
+                    if !matches!(token, Tok::Comment(_)) {
+                        line_is_comment_only = false;
+                    }
```

Comment:
I could rename `line_is_comment_only` to `line_is_trivia`. However, since new lines and indents are skipped, a line can have `line_is_comment_only` set to `true` if and only if it contains a comment. Therefore I think it is more helpful to have `line_is_comment_only` as the variable name.